refactor: improve docs, comments, tests, and small refactors

JimLarson opened the pull request:
## Description

Improve the documentation and comments to enable local auditing of all code.

Incidental small refactoring and test expansion.

Test coverage could be improved, but not in this PR.

### Author Checklist

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [x] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [x] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [x] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [x] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

JimLarson committed:
refactor: improve docs, comments, tests, and small refactors

gibson042 commented on x/auth/vesting/cmd/vestcalc/README.md:
```diff
@@ -32,20 +32,36 @@ tool for other options.
-- `--months`: The number of months to vest over.
-- `--time`: The time of day of the vesting event, in 24-hour HH:MM format.
-  Defaults to midnight.
+- `"start_time"`: integer UNIX time;
```
```suggestion
- `"start_time"`: integer seconds since the POSIX epoch;
```

JimLarson replied:
I'll stick with "UNIX time" for the concept, since a) it matches the the Go `time` package API (which doesn't mention an epoch, only "UNIX time"), and b) "UNX time" has its own wikipedia page - "POSIX time" redirects there.

gibson042 commented on x/auth/vesting/cmd/vestcalc/README.md:
```diff
@@ -32,20 +32,36 @@ tool for other options.
-stdout. The following flags control the output:
+When the `--write` flag is set, the tool will write a schedule to stdout.
+The format is the the format expected by the `create-periodic-vesting-account`
+or `create-clawback-vesting-account` cli commands, namely a JSON
```
```suggestion
or `create-clawback-vesting-account` cli commands, namely JSON text representing an
```

JimLarson replied:
JSON implies text. However, from all the format-related comments, it could be confusing to have it spread out. I've added a single format section with an example that I hope makes it more clear.

gibson042 commented on x/auth/vesting/cmd/vestcalc/README.md:
```diff
@@ -32,20 +32,36 @@ tool for other options.
-- `--time`: The time of day of the vesting event, in 24-hour HH:MM format.
-  Defaults to midnight.
+- `"start_time"`: integer UNIX time;
+- `"periods"`: a JSON array of JSON objects with members:
```
I'm assuming there's not another layer of serialization here (e.g., output looks like `{ …, "periods": [ { "coins": "…", "length_seconds": <digits> }, … ] }`).
```suggestion
- `"periods"`: an array of objects with members:
```

JimLarson replied:
Correct - I've added an example and moved this up to a "format" section, where the JSON context is more clear and I can drop the superfluous and confusing repeats.

gibson042 commented on x/auth/vesting/cmd/vestcalc/README.md:
```diff
@@ -32,20 +32,36 @@ tool for other options.
-  Defaults to midnight.
+- `"start_time"`: integer UNIX time;
+- `"periods"`: a JSON array of JSON objects with members:
+    - "coins": string giving the text coins format for the amount vested at this event;
```
I think example output would help, especially for this field. Probably just a URL pointing to something that already exists.

JimLarson replied:
I don't see a URL for the coins text format. I'll add a multi-denom example.

gibson042 commented on x/auth/vesting/cmd/vestcalc/README.md:
```diff
@@ -32,20 +32,36 @@ tool for other options.
+The following flags control the output:
+
+- `--coins:` The total coins to vest, e.g. `100ubld,50urun`.
+- `--months`: The number of monthly to vesting events.
```
I think there's a typo here, but the explanation could still be a little more clear.
```suggestion
- `--months`: The total number of months to complete vesting.
```

JimLarson replied:
Done.

gibson042 commented on x/auth/vesting/cmd/vestcalc/README.md:
```diff
@@ -32,20 +32,36 @@ tool for other options.
   next month. Specified in the format `YYYY-MM-DD` or `YYYY-MM-DDThh:mm`,
   e.g. `2006-01-02T15:04` for 3:04pm on January 2, 2006.
-- `--cliffs`: One or more vesting cliffs in `YYYY-MM-DD` or `YYYY-MM-DDThh:mm`
+- `--time`: The time of day of the vesting events, in 24-hour HH:MM format.
```
`--start` and `--time` are interpreted as UTC, right? And does `--time` override hours and minutes in `--start`?

JimLarson replied:
No. There's a note above that times are interpreted in the local timezone. I added a repeat of that comment here to clarify. The time-of-day of the start time has no bearing on the time-of-day of the later vesting events.

gibson042 commented on x/auth/vesting/cmd/vestcalc/README.md:
```diff
@@ -32,20 +32,36 @@ tool for other options.
+day of the month (or the last day of the month, if the month does not have a
+sufficient number of days), for the specified number of months. The total coins
+to vest will be divided as evenly as possible among all the vesting events.
+Lastly, all events before the last cliff time, if any, are consolidated into a single even
```
```suggestion
Lastly, all events before the last cliff time, if any, are skipped and consolidated into a single event
```

JimLarson replied:
I think "consolidated" is sufficient.

gibson042 commented on x/auth/vesting/cmd/vestcalc/vestcalc.go:
```diff
@@ -81,6 +98,7 @@ func divideCoins(coins sdk.Coins, divisor int) ([]sdk.Coins, error) {
 // monthlyVestTimes generates timestamps for successive months after startTime.
 // The monthly events occur at the given time of day. If the month is not
 // long enough for the desired date, the last day of the month is used.
+// The number of months must be postive.
```
Oh nice, CI caught this typo!

JimLarson replied:
Fixed.

gibson042 commented on x/auth/vesting/cmd/vestcalc/vestcalc.go:
```diff
@@ -431,10 +453,10 @@ var (
-// readCmd reads periods in JSON from stdin and writes a sequence of vesting
-// events in local time to stdout.
+// readCmd reads a schedule file from stdin and writes a sequence of vesting
+// events in local time to stdout. See README.md for the format.
```
In _local_ time? Oof. If that's the case we should probably find a way to convey UTC offsets.

JimLarson replied:
I think it matches the anticipated usage pattern. The comment about timezones in `README.md` describes how to get other timezones.

gibson042 commented on x/auth/vesting/cmd/vestcalc/vestcalc.go:
```diff
@@ -13,6 +13,9 @@ import (
 	"github.com/cosmos/cosmos-sdk/x/auth/vesting/client/cli"
 )
 
+// vestcalc is a utility for creating or reading period files
```
Vocabulary consistency nit.
```suggestion
// vestcalc is a utility for creating or reading schedule files
```

JimLarson replied:
Done.

gibson042 reviewed (approved):
Thanks!

JimLarson committed:
docs: review feedback

JimLarson commented:
I've addressed your feedback, @gibson042 , PTAL. I went along with most suggestions, either by letter or spirit. If you feel strongly about any changes I resisted, please re-raise them.
